subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -199,6 +202,8 @@ public void validateMutation(MutationType type) {
     private final ImmutableAttributesFactory attributesFactory;
     private final FileCollection intrinsicFiles;
 
+    private final ProjectStateRegistry projectStateRegistry;
+
     private final DisplayName displayName;
 
     public DefaultConfiguration(DomainObjectContext domainObjectContext,
```

Comment:
Does `markReferencedProjectConfigurationsObserved` look like it's doing something suspicious? It looks like it was doing work under the old lock (maybe) and may be cross project boundaries?

Reply:
It wasn't using the old lock - this is more internal state synchronization.  This one is a little weird because the synchronization only affects this single variable, so it seems a bit heavy to spin the project mutation lock here, but it's probably the right thing to do (this only comes in to play when `projectA.configurations.foo extendsFrom projectB.configurations.bar`).  Even better would be to find some way to just eliminate this cross-project behavior, but I couldn't come up with anything that wouldn't be major surgery.